src/passes/CodePushing.cpp

```diff
@@ -124,11 +124,26 @@ class Pusher {
       return nullptr;
     }
     auto index = set->index;
-    // to be pushable, this must be SFA and the right # of gets,
-    // but also have no side effects, as it may not execute if pushed.
+    // To be pushable, this must be SFA and the right # of gets.
+    //
+    // It must also not have side effects, as it may no longer execute after it
+    // is pushed, since it may be behind a condition that ends up false some of
+    // the time. However, removable side effects are ok here. The general
+    // problem with removable effects is that we can only remove them, but not
+    // move them, because of stuff like this:
+    //
+    //   if (x != 0) foo(1 / x);
+    //
+    // If we move 1 / x to execute unconditionally then it may trap, but it
+    // would be fine to remove it. This pass does not move code to places where
+    // it might execute more, but *less*: we keep the code behind any conditions
+    // it was already behind, and potentially put it behind further ones. In
+    // effect, we "partially remove" the code, making it not execute some of the
+    // time, which is fine.
```

Comment:
Code LGTM, but not sure if I understand this... in `trapsNeverHappen` mode, we simply assume traps never happen, so it doesn't matter if we execute possibly trapping code more of less.. no?

Reply:
It's a little tricky. Given `if (x != 0) foo(1/x)` we assume it never traps as it executes normally. But the only reason it doesn't trap might be that the `if` prevents it (that's what "executes normally" means, sort of). So moving the `1/x` out of the if is risky. (But in this pass we only execute code less, not more, so it's ok.)

The old "ignore implicit traps" option was basically "nothing can trap (even if you move it)", but the movement turned out to be a problem, so we added trapsNeverHappen which is more subtle.

Docs for this should be here: (but it is just a tricky topic, unfortunately - more complex than I'd like)

https://github.com/WebAssembly/binaryen/blob/79e032ac1124f82729bb21684f370044612c7124/src/pass.h#L119-L167

Reply:
Thanks for the pointer. I had read it before but apparently forgot it... 😅